posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md

```diff
@@ -0,0 +1,209 @@
+
+With the GitHub Action workflows available for Linux, it seemed reasonable that we would like to test the same on macOS. This [issue](https://github.com/rgommers/scipy/issues/64) by [Rohit Goswami](https://github.com/HaoZeke), where he documents the problems with the macOS build, made sense that we can implement a workflow with little to almost no tweaks. However, I wanted to implement the conda-based approach rather than putting forward a native build.
+
+GitHub Actions allows us to perform macOS builds quite generously, and creating a workflow for the same was relatively easy. After this, we had to make a couple of iterations to verify if the compiler caching exactly works and how performant it is compared to a fresh build. I also implemented conda caching for faster installs using [Mamba](https://github.com/mamba-org/mamba), which reduced the effective build time by 30%.
+
+[PR #65](https://github.com/rgommers/scipy/pull/65) containing the commits for the code required to set up the GitHub Action workflow for macOS with compiler caching and conda caching.
+
+## Automating CI benchmarking on GitHub Actions
+
+This task was heavily influenced by the work done by [Jaime Rodríguez-Guerra](https://github.com/jaimergp) on [Scikit-Image](https://github.com/scikit-image/scikit-image)  [PR #5424](https://github.com/scikit-image/scikit-image/pull/5424). We wanted the benchmarks to be run automatically on specific Pull Requests with apt labels or during releases. It, however had a few problems that we wanted to address before moving forward:
+
+- Benchmarking requires to happen on specific hardware under specific conditions, which is difficult to reproduce.
+
+- Benchmarking on CI is even harder, primarily when different runners are provided with shared resources.
+
+- Having dedicated hardware for the same can lead to abuse and unnecessary noise.
+
+We decided to use either [Cachegrind](https://valgrind.org/docs/manual/cg-manual.html) or [Jaime’s approach](https://labs.quansight.org/blog/2021/08/github-actions-benchmarks/) to tackle all of these problems, which talks about relative benchmarking on GitHub Actions. Both scikit-image and SciPy use Airspeed Velocity (`asv`) for benchmarking. This rules out the use of cachegrind since `asv` is not compatible with the same.
+
+Following in Jaime’s footsteps, I used `asv`’s subcommand called `continuous` to run a benchmark test on two corresponding commits. This will collect all the statistics together and give us a log and results, which can be uploaded as an artifact after the workflow gets over. It would allow the maintainers to directly download the same and verify on their local machine before moving ahead with a pull request or a release.
+
+[PR #67](https://github.com/rgommers/scipy/pull/67) contains the commits for the code required to set up the GitHub Action workflow for automated CI benchmarking using `asv` and artifact upload.
+
+## Strategizing the CI Matrix for SciPy
```

Comment:
You could consider using a screenshot of the CI matrix, if you remove the yellow lines and make it a bit more compact, this could look nice/useful:

<img width="781" alt="image" src="https://user-images.githubusercontent.com/98330/136621128-425a6093-2f77-4c78-b580-601c60dde5cb.png">